File: src/lib/wowhead-client/item-armor.ts
import {
  fetchWithCache, getLatestExpansionHavingUrl, getZamBaseUrl, ZamUrl,
} from './zam-url';

// Slots enum used by orchestration and attachments
export enum EquipmentSlot {
  Head = 1,
  Shoulder = 3,
  Shirt = 4,
  Chest = 5,
  Waist = 6,
  Legs = 7,
  Feet = 8,
  Wrist = 9,
  Hands = 10,
  MainHand = 12,
  OffHand = 13,
  Shield = 14,
  Ranged = 15,
  Cloak = 16,
  Tabard = 19,
  Robe = 20,
  Holdable = 23,
  RangedRight = 26,
}

interface ItemFile {
  FileDataId: number;
  Race: number;
  Gender: number;
  Class: number;
  ExtraData: number;
}

export interface ItemData {
  Textures: { [k: string]: number } | null;
  Textures2: { [k: string]: number } | null;
  ModelFiles: { [modelId: string]: ItemFile[] };
  TextureFiles: { [textureId: string]: ItemFile[] };
  Item: {
    Flags: number;
    InventoryType: number;
    ItemClass: number;
    ItemSubClass: number;
    GeosetGroup: number[];
    AttachGeosetGroup?: number[];
    GeosetGroupOverride?: number[];
    HideGeosetMale?: { RaceId: number; GeosetGroup: number; RaceBitSelection: number }[];
    HideGeosetFemale?: { RaceId: number; GeosetGroup: number; RaceBitSelection: number }[];
  };
  ComponentModels: { [componentId: string]: number };
  ComponentTextures: { [componentId: string]: number };
}

const debug = false;

const ArmorSlots = [
  EquipmentSlot.Head,
  EquipmentSlot.Shoulder,
  EquipmentSlot.Shirt,
  EquipmentSlot.Chest,
  EquipmentSlot.Waist,
  EquipmentSlot.Legs,
  EquipmentSlot.Feet,
  EquipmentSlot.Wrist,
  EquipmentSlot.Hands,
  EquipmentSlot.Cloak,
  EquipmentSlot.Tabard,
  EquipmentSlot.Robe,
];

const slotBackup = {
  5: 20,
};

function itemMetaCandidatePaths(displayId: number, preferredSlot: number | null): string[] {
  const paths: string[] = [];
  const seen = new Set<number>();
  const addSlot = (slot: number) => {
    if (slot <= 0 || seen.has(slot)) return;
    seen.add(slot);
    paths.push(`meta/armor/${slot}/${displayId}.json`);
  };
  if (preferredSlot) {
    addSlot(preferredSlot);
    const backup = slotBackup[preferredSlot as keyof typeof slotBackup];
    if (backup) addSlot(backup);
  }
  for (const slot of ArmorSlots) {
    addSlot(slot);
  }
  paths.push(`meta/item/${displayId}.json`);
  return paths;
}

async function fetchItemMetaAtPath(expansion: ZamUrl['expansion'], path: string): Promise<ItemData> {
  let exp = expansion;
  if (exp === 'latest-available') {
    exp = await getLatestExpansionHavingUrl(path);
  }
  const url = `${getZamBaseUrl(exp)}/${path}`;
  try {
    const res = await fetchWithCache(url);
    return JSON.parse(res) as unknown as ItemData;
  } catch (e) {
    if (expansion !== 'latest-available') {
      try {
        const fallback = await getLatestExpansionHavingUrl(path);
        if (fallback !== exp) {
          const fallbackUrl = `${getZamBaseUrl(fallback)}/${path}`;
          const res = await fetchWithCache(fallbackUrl);
          return JSON.parse(res) as unknown as ItemData;
        }
      } catch {
        // fall through
      }
    }
    throw e;
  }
}

export async function fetchItemMeta(zam: ZamUrl): Promise<ItemData> {
  if (zam.type !== 'item') throw new Error('fetchItemMeta expects a ZamUrl of type item');
  let slotId: number | null = zam.slotId;
  if (slotId && !ArmorSlots.includes(slotId)) slotId = null;

  let lastError: unknown;
  for (const path of itemMetaCandidatePaths(zam.displayId, slotId)) {
    try {
      return await fetchItemMetaAtPath(zam.expansion, path);
    } catch (e) {
      lastError = e;
    }
  }
  throw lastError;
}

File: golang/internal/wowhead/item_armor.go
package wowhead

import (
	"encoding/json"
	"fmt"
)

// EquipmentSlot identifies wow equipment slots.
type EquipmentSlot int

const (
	SlotHead EquipmentSlot = 1
	SlotShoulder EquipmentSlot = 3
	SlotShirt EquipmentSlot = 4
	SlotChest EquipmentSlot = 5
	SlotWaist EquipmentSlot = 6
	SlotLegs EquipmentSlot = 7
	SlotFeet EquipmentSlot = 8
	SlotWrist EquipmentSlot = 9
	SlotHands EquipmentSlot = 10
	SlotMainHand EquipmentSlot = 12
	SlotOffHand EquipmentSlot = 13
	SlotShield EquipmentSlot = 14
	SlotRanged EquipmentSlot = 15
	SlotCloak EquipmentSlot = 16
	SlotTabard EquipmentSlot = 19
	SlotRobe EquipmentSlot = 20
	SlotHoldable EquipmentSlot = 23
	SlotRangedRight EquipmentSlot = 26
)

var armorSlots = []EquipmentSlot{
	SlotHead, SlotShoulder, SlotShirt, SlotChest, SlotWaist, SlotLegs, SlotFeet,
	SlotWrist, SlotHands, SlotCloak, SlotTabard, SlotRobe,
}

var slotBackup = map[int]int{5: 20}

// ItemFile is a race/gender/class keyed file entry.
type ItemFile struct {
	FileDataID int `json:"FileDataId"`
	Race       int `json:"Race"`
	Gender     int `json:"Gender"`
	Class      int `json:"Class"`
	ExtraData  int `json:"ExtraData"`
}

// ItemData is zam item meta JSON.
type ItemData struct {
	Textures         map[string]int            `json:"Textures"`
	Textures2        map[string]int            `json:"Textures2"`
	ModelFiles       map[string][]ItemFile     `json:"ModelFiles"`
	TextureFiles     map[string][]ItemFile     `json:"TextureFiles"`
	ComponentModels  map[string]int            `json:"ComponentModels"`
	ComponentTextures map[string]int           `json:"ComponentTextures"`
	Item             ItemInfo                    `json:"Item"`
}

// ItemInfo holds item metadata fields.
type ItemInfo struct {
	Flags            int `json:"Flags"`
	InventoryType    int `json:"InventoryType"`
	ItemClass        int `json:"ItemClass"`
	ItemSubClass     int `json:"ItemSubClass"`
	GeosetGroup      []int `json:"GeosetGroup"`
	HideGeosetMale   []HideGeosetEntry `json:"HideGeosetMale"`
	HideGeosetFemale []HideGeosetEntry `json:"HideGeosetFemale"`
}

// HideGeosetEntry is a hide-geoset rule for a race.
type HideGeosetEntry struct {
	RaceID      int `json:"RaceId"`
	GeosetGroup int `json:"GeosetGroup"`
}

func itemMetaCandidatePaths(displayID, preferredSlot int) []string {
	paths := make([]string, 0, len(armorSlots)+2)
	seen := map[int]bool{}
	addSlot := func(slot int) {
		if slot <= 0 || seen[slot] {
			return
		}
		seen[slot] = true
		paths = append(paths, fmt.Sprintf("meta/armor/%d/%d.json", slot, displayID))
	}
	addSlot(preferredSlot)
	if backup, ok := slotBackup[preferredSlot]; ok {
		addSlot(backup)
	}
	for _, slot := range armorSlots {
		addSlot(int(slot))
	}
	paths = append(paths, fmt.Sprintf("meta/item/%d.json", displayID))
	return paths
}

func fetchItemMetaAtPath(client *HTTPClient, expansion Expansion, path string) (ItemData, error) {
	exp := expansion
	if exp == ExpansionLatestAvailable {
		var err error
		exp, err = GetLatestExpansionHavingURL(client, path)
		if err != nil {
			return ItemData{}, err
		}
	}
	url := GetZamBaseURL(exp) + "/" + path
	text, err := FetchWithCache(client, url)
	if err != nil && expansion != ExpansionLatestAvailable {
		if fallback, err2 := GetLatestExpansionHavingURL(client, path); err2 == nil && fallback != exp {
			url = GetZamBaseURL(fallback) + "/" + path
			text, err = FetchWithCache(client, url)
		}
	}
	if err != nil {
		return ItemData{}, err
	}
	var data ItemData
	if err := json.Unmarshal([]byte(text), &data); err != nil {
		return ItemData{}, err
	}
	return data, nil
}

// FetchItemMeta fetches item or armor meta JSON from zam modelviewer.
func FetchItemMeta(client *HTTPClient, expansion Expansion, displayID int, slotID *int) (ItemData, error) {
	preferredSlot := 0
	if slotID != nil {
		preferredSlot = *slotID
		if preferredSlot > 0 && !containsArmorSlot(EquipmentSlot(preferredSlot)) {
			preferredSlot = 0
		}
	}
	var lastErr error
	for _, path := range itemMetaCandidatePaths(displayID, preferredSlot) {
		data, err := fetchItemMetaAtPath(client, expansion, path)
		if err == nil {
			return data, nil
		}
		lastErr = err
	}
	if lastErr != nil {
		return ItemData{}, lastErr
	}
	return ItemData{}, fmt.Errorf("item meta not found for display %d", displayID)
}

func containsArmorSlot(slot EquipmentSlot) bool {
	for _, s := range armorSlots {
		if s == slot {
			return true
		}
	}
	return false
}
